Add the persistent-selection extension

ankon opened the pull request:
### Description

Fixes #601 by adding a default extension to decorate the current selection, IFF the `persistentSelectionClass` builtin configuration is provided. 

### Checklist

- [x] I have read the [**contributing**](https://github.com/remirror/remirror/blob/HEAD/docs/contributing.md) document.
- [x] My code follows the code style of this project and `pnpm fix` completed successfully.
- [ ] I have updated the documentation where necessary.
- [x] New code is unit tested and all current tests pass when running `pnpm test`.

ankon commented:
Two open questions:

1. Documentation: Sure, where? :)
2. Changeset: Uuuh. Do I do that, or do you do that? I suspect it involves running `pnpm changeset`.

ankon commented:
> Changeset: Uuuh. Do I do that, or do you do that? I suspect it involves running pnpm changeset.

Right, I think I understand -- changeset == documentation piece. I added something, and I think this is a "patch" change. If the user doesn't configure the extension, then nothing will change.

ifiokjr commented on .changeset/unlucky-pets-melt.md:
```diff
@@ -0,0 +1,9 @@
+---
+'@remirror/core': patch
```
Since this is a new feature, make it a minor change.

```suggestion
'@remirror/core': minor
```

ifiokjr commented on packages/@remirror/core/src/builtins/__tests__/persistent-selection-extension.spec.ts:
```diff
@@ -0,0 +1,52 @@
+
+  selectText(4);
+
+  setTimeout(() => {
```
Not sure that we need the `setTimeout` here. Updates are immediately available. 

Also, the test seems to be focused on the implementation details. IMO the important things to test are:

- The current selection is wrapped with an element that has the class `persistent-selection`. So `expect(view.dom.innerHTML).toMatchInlineSnapshot()` is a good enough test.
- When focus is lost the selection class is maintained. I think this was the main goal of the original issue.

ankon replied:
I had to add the timeout, neither `act()` nor simply chaining the calls passed the tests.

ankon replied:
> When focus is lost the selection class is maintained.

Added a test for this case as well now.

ifiokjr commented on packages/@remirror/core/src/builtins/__tests__/persistent-selection-extension.spec.ts:
```diff
@@ -0,0 +1,52 @@
+
+  add(doc(p('Some text')));
+
+  selectText(4);
```
Selection should probably be a range here. `selectText(4, 9)`. It raises a good question. Do we want the selection to even be wrapped when empty?

ifiokjr replied:
@ronnyroeller, any thoughts?

ankon replied:
`selectText` defaults the end to the end of the text, but I can make this explicit of course.

ifiokjr commented on packages/@remirror/core/src/builtins/persistent-selection-extension.ts:
```diff
@@ -0,0 +1,50 @@
+// XXX: @remirror/core exports a _type_ 'DecorationSet', but we need access to the static methods as well.
```
I'm not sure what this comment means. Do you want to export the `DecorationSet` from `@remirror/core`?

ankon replied:
`@remirror/Core` exports the `DecorationSet` and `Decoration` types, but using `DecorationSet.empty` lead to Compiler errors as that would need the static property. 

Arguably: I can remove the comment if importing from `@remirror/pm/view` is ok.

ifiokjr replied:
Yeah, importing from `@remirror/pm/view` is expected. The types exported by `@remirror/core-types` are all aliases. Initially, it was to reduce the number of import statements needed when creating an extension. The types for `EditorView`, `EditorSchema`, `Selection`, `ProsemirrorNode`, `Mark` are used so often that it made sense to alias them.

ifiokjr commented on packages/@remirror/core/src/builtins/persistent-selection-extension.ts:
```diff
@@ -0,0 +1,50 @@
+import type { CreatePluginReturn } from '../types';
+
+export interface PersistentSelectionOptions {
+  persistentSelectionClass?: Static<string>;
```
Good use of `Static` 👍

ifiokjr commented on packages/@remirror/core/src/builtins/persistent-selection-extension.ts:
```diff
@@ -0,0 +1,50 @@
+            return DecorationSet.empty;
+          }
+
+          if (selection.empty) {
```
Now I'm not even sure how the previous test passes 😄 
It looks like you return early for empty selections.

ifiokjr replied:
Oh I see, it passes because it's never run. The `setTimeout` means that the error will be thrown after the test.

ifiokjr commented:
Ooops I submitted the review too early last time without comment 😢  

Here's the reason for requesting changes.

> The editor should give the illusion that the current selection stays even when the editor lost focus

I think this part of issue #601 hasn't really been resolved. There is a selection added to the editor, however, it's not been related to the whether the editor is  in focus. Unfortunately, docs don't build for PR's from an outside fork, so I can't try out your code in the playground without building it myself. 

I've added @ronnyroeller as a reviewer as well.

ifiokjr commented on packages/@remirror/core/src/builtins/persistent-selection-extension.ts:
```diff
@@ -0,0 +1,50 @@
+
+@extensionDecorator<PersistentSelectionOptions>({
+  defaultOptions: {
+    persistentSelectionClass: undefined,
```
I actually feel like there should a default value here. Something like `selected-text`.

ankon replied:
I initially had a default, but that will break compatibility (as many of the unit tests show). 

Of course, the question is whether 100% compatibility is relevant here in the context of remirror.

ifiokjr replied:
Oh, that's a valid point. 

I don't mind updating the snapshots, though. It's a core piece of functionality and my feeling is that it will be an expected behaviour for the majority of users. Also, the class is pretty harmless by itself, unless styles are assigned.

ifiokjr commented on .changeset/unlucky-pets-melt.md:
```diff
@@ -0,0 +1,9 @@
+'@remirror/core': patch
+---
+
+Optionally allow to style the currently selected text
```
Nice use of changesets here 👍

ankon commented:
As to addressing #601: what I found is that the selection doesn't get lost at all, the problem is merely a display issue of not being able to _see_ the selection. 

For that reason this extension/plugin is fairly trivial, it just needs to keep a decoration on the selected area: as soon as the browser removes whatever it does by default the decoration takes over, thereby keeping things visible.

ankon commented:
Moved it back to draft -- I think I see the issue with the tests, so I'll start from there and then update the branch when I have tests that actually test the right thing.

ankon commented:
Alright, next update, addressing most of the feedback.

> Oh I see, it passes because it's never run. The `setTimeout` means that the error will be thrown after the test.

Indeed, the tests actually didn't do anything sensible. The problem here was actually that looking at the `props.decorations` was pointless (it contains incoming properties, and not a view on the _current_ properties as I wrongly assumed). The new tests simply use Jest's snapshots, which works just fine to show the extension getting activated.

I still need to look into writing tests that show that a change in focus away from the component still leaves these decorations around.

One big issue though is the second commit in this set, where I don't really know the right direction forward. The code-block extension creates simple `<span>text</span>` decorations for `text`-typed refractor nodes, and when the persistent-selection extension adds a decoration on exactly one of these we end up with a calcuated `class` of "persistent-selection " (note the trailing space). prosemirror-view then tries to parse that value into tokens, so that it can maintain the `classList` DOMTokenList manually: https://github.com/ProseMirror/prosemirror-view/blob/b9fd1bc695a9c69ac3c472ffdbd61115b51d0588/src/viewdesc.js#L962
The problem in prosemirror-view is that it may produce empty tokens, which are explicitly illegal in calls to `DOMTokenList.add` or `DOMTokenList.remove` (see https://dom.spec.whatwg.org/#dom-domtokenlist-add). I'm thinking that prosemirror-view is to blame here, and that it should deduplicate/remove excess whitespace (as MDN hints here: https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList#Trimming_of_whitespace_and_removal_of_duplicates). But at the same time it also looks weird to produce decorations without a class, so I added a 'text' class in the code-block extension, which _also_ fixes the problem by avoiding the prosemirror-view issue. WDYT?

ifiokjr commented:
Okay, thanks for the write-up, although I'm not sure I fully understand. Is there an error being produced?

ankon commented:
The error is really looking innocent enough when you only apply the first commit:
~~~~
  ● plugin › Backspace › avoids stepping into the previous node with non empty selection

    SyntaxError: The token provided must not be empty.
~~~~

The call stack of that error is hidden, and looks like this:
~~~~
validateTokens (.../remirror/node_modules/.pnpm/jsdom@16.4.0/node_modules/jsdom/lib/jsdom/living/nodes/DOMTokenList-impl.js:13)
add (.../remirror/node_modules/.pnpm/jsdom@16.4.0/node_modules/jsdom/lib/jsdom/living/nodes/DOMTokenList-impl.js:110)
add (.../remirror/node_modules/.pnpm/jsdom@16.4.0/node_modules/jsdom/lib/jsdom/living/generated/DOMTokenList.js:159)
patchAttributes (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:967)
patchOuterDeco (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:949)
applyOuterDeco (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:981)
create (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:597)
addNode (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:1131)
(anonymous function) (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:656)
iterDeco (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:1249)
updateChildren (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:639)
updateInner (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:725)
update (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:717)
updateNextNode (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:1116)
(anonymous function) (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:654)
iterDeco (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:1197)
updateChildren (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:639)
updateInner (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:725)
update (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/viewdesc.js:717)
updateStateInner (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/index.js:160)
updateState (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/index.js:114)
updateState (.../remirror/packages/@remirror/dom/src/dom.ts:114)
(anonymous function) (.../remirror/packages/@remirror/core/src/editor-wrapper.ts:340)
dispatch (.../remirror/node_modules/.pnpm/prosemirror-view@1.15.5/node_modules/prosemirror-view/src/index.js:367)
dispatchTextSelection (.../remirror/packages/jest-prosemirror/src/jest-prosemirror-editor.ts:188)
(anonymous function) (.../remirror/packages/jest-remirror/src/jest-remirror-editor.ts:309)
(anonymous function) (.../remirror/packages/@remirror/extension-code-block/src/__tests__/code-block-extension.spec.ts:147)
Promise.then.completed (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/utils.js:230)
callAsyncCircusFn (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/utils.js:170)
_callCircusTest (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:212)
[ async function ]
_callCircusTest (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:197)
_runTest (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:149)
[ async function ]
_runTest (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:106)
_runTestsForDescribeBlock (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:63)
[ async function ]
_runTestsForDescribeBlock (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:36)
_runTestsForDescribeBlock (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:57)
[ async function ]
_runTestsForDescribeBlock (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:36)
_runTestsForDescribeBlock (.../remirror/node_modules/.pnpm/jest-circus@26.4.2/node_modules/jest-circus/build/run.js:57)
~~~~

The problem occurs in the prosemirror-view `ViewDesc.create` call, where it needs to apply 2 decorations:
```json
[
  {
    "from": 0,
    "to": 4,
    "type": {
      "spec": {},
      "attrs": {
        "class": "persistent-selection"
      }
    }
  },
  {
    "from": 0,
    "to": 4,
    "type": {
      "spec": {},
      "attrs": {
        "class": ""
      }
    }
  }
]
```

The `patchAttributes` call is trying to update the `class` attribute with the value "persistent-state " (trailing space!), when then fails when it tries to add the "new" empty token to the `classList`.

ankon commented:
The TL;DR of the question is probably: Are you ok with another PR that would change the code-block extension to set a (configurable) class on `text` nodes, and that would otherwise not create annotations for `text` nodes?

ifiokjr commented:
> The TL;DR of the question is probably: Are you ok with another PR that would change the code-block extension to set a (configurable) class on `text` nodes, and that would otherwise not create annotations for `text` nodes?

I'm okay with that. Meanwhile, we could just ignore the code block, via code. If the parent node is a code block, don't create a decoration, if that's easier.

ankon commented:
> I'm okay with that. Meanwhile, we could just ignore the code block, via code. If the parent node is a code block, don't create a decoration, if that's easier.

I did quite some back-and-forth here, but ultimately I think the approach of one the one hand preventing to create decorations without a `class`, and on the hand still retaining the ability to have all parts of a code block targetable through CSS, is the best way forward here. I'm will still create an issue for prosemirror-view as well, as the situation seems somewhat "improveable" :)

See https://github.com/remirror/remirror/pull/630.

**For now I kept this PR on top of #630, as otherwise the tests will fail. I will rebase onto the `next` branch before marking this "ready for review"**

ifiokjr commented:
Once you've updated this branch, I'll merge in your work.

ankon commented:
Thank you for your patience! I think from my side I covered now everything from the review comments as well, so "should be good to go" :)

ifiokjr commented on packages/@remirror/core/src/builtins/__tests__/persistent-selection-extension.spec.ts:
```diff
@@ -0,0 +1,60 @@
+extensionValidityTest(PersistentSelectionExtension);
+
+describe('configuration', () => {
+  [
```
Jest actually has it's own builtin `.each` test method. 

https://jestjs.io/docs/en/api#testeachtablename-fn-timeout

ankon replied:
I am starting to like jest :)

ifiokjr commented on packages/@remirror/core/src/builtins/__tests__/persistent-selection-extension.spec.ts:
```diff
@@ -0,0 +1,60 @@
+});
+
+describe('focus', () => {
+  test('retains selection and decoration', () => {
```
Nitpick, I prefer `test()` when it's at top level and `it()` when nested in a `describe` block. 

I don't even know why, and you can leave it as is.

ifiokjr commented on packages/@remirror/core/src/builtins/__tests__/persistent-selection-extension.spec.ts:
```diff
@@ -0,0 +1,60 @@
+    expect(editor.state.selection.from).toBe(range.from);
+    expect(editor.state.selection.to).toBe(range.to);
+
+    // Add a new element and focus that one
```
You can also use `editor.view.dom.blur()`.

ankon replied:
I needed to explicitly cast the `dom` property to `HTMLElement` to have access to `blur()`, but this definitely reads better and makes it more obvious what the point of the test is.

ifiokjr replied:
Oh interesting, try `editor.dom`, just realised that might already be cast to an HTML element.

ifiokjr replied:
Ignore that, it isn't. I'll update it so that it is.

ifiokjr commented on .changeset/unlucky-pets-melt.md:
```diff
@@ -0,0 +1,9 @@
+
+This adds a new option for the builtin preset, `persistentSelectionClass`. If that is set to a valid CSS class name any selection in the editor will be decorated with this class.
+
+This can be used to keep a indication for the current selection even when the focus changes away from the editor.
```
```suggestion
This can be used to keep an indication for the current selection even when the focus changes away from the editor.
```

ifiokjr reviewed (approved):
The tests are looking much better 🥳 

Just some super, super minor tweaks which you can feel free to ignore. I'll merge in a few hours regardless.
